src/equations/shallow_water_2d.jl

```diff
@@ -0,0 +1,535 @@
+# By default, Julia/LLVM does not use fused multiply-add operations (FMAs).
+# Since these FMAs can increase the performance of many numerical algorithms,
+# we need to opt-in explicitly.
+# See https://ranocha.de/blog/Optimizing_EC_Trixi for further details.
+@muladd begin
+
+
+@doc raw"""
+    ShallowWaterEquations2D(gravity, H0)
+
+Shallow water equations (SWE) in two space dimensions. The equations are given by
+```math
+\begin{aligned}
+  \frac{\partial h}{\partial t} + \frac{\partial}{\partial x}(h v_1)
+    + \frac{\partial}{\partial y}(h v_2) &= 0 \\
+    \frac{\partial}{\partial t}(h v_1) + \frac{\partial}{\partial x}\left(h v_1^2 + \frac{g}{2}h^2\right)
+    + \frac{\partial}{\partial y}(h v_1 v_2) + g h \frac{\partial b}{\partial x} &= 0 \\
+    \frac{\partial}{\partial t}(h v_2) + \frac{\partial}{\partial x}(h v_1 v_2)
+    + \frac{\partial}{\partial y}\left(h v_2^2 + \frac{g}{2}h^2\right) + g h \frac{\partial b}{\partial y} &= 0.
+\end{aligned}
+```
+The unknown quantities of the SWE are the water height ``h`` and the velocities ``\mathbf{v} = (v_1, v_2)^T``.
+The gravitational constant is denoted by `g` and the (possibly) variable bottom topography function ``b(x,y)``.
+Conservative variable water height ``h`` is measured from the bottom topography ``b``, therefore one
+also defines the total water height as ``H = h + b``.
+
+The additional quantity ``H_0`` is also available to store a reference value for the total water height that
+is useful to set initial conditions or test the "lake-at-rest" well-balancedness.
+
+The bottom topography function ``b(x,y)`` is set inside the initial condition routine
+for a particular problem setup. To test the conservative form of the SWE one can set the bottom topography
+variable `b` to zero.
+
+In addition to the unknowns, Trixi currently stores the bottom topography values at the approximation points
+despite being fixed in time. This is done for convenience of computing the bottom topography gradients
+on the fly during the approximation as well as computing auxiliary quantities like the total water height ``H``
+or the entropy variables.
+This affects the implementation and use of these equations in various ways:
+* The flux values corresponding to the bottom topography must be zero.
+* The bottom topography values must be included when defining initial conditions, boundary conditions or
+  source terms.
+* [`AnalysisCallback`](@ref) analyzes this variable.
+* Trixi's visualization tools will visualize the bottom topography by default.
```

Comment:
What happens with `flux_lax_friedrichs` and `flux_hll`? I would expect that we need to specialize the corresponding dissipation terms as we do for acoustics, see https://github.com/trixi-framework/Trixi.jl/blob/e5b237e1a123f732e2c6c6203780ea9571cbdb5b/src/equations/acoustic_perturbation_2d.jl#L423-L430
Otherwise, I would expect to see some changes of the bottom topography over time. Does something like that happen, @andrewwinters5000?

Reply:
Yes, this was a problem for the APE :+1: and fixed by @lchristm

Reply:
I hadn't noticed anything but the final time for my simulations is much shorter than the APE `tspan`. Thus, maybe the bottom values are changing but very slowly. So for the sake of consistency as well as to be on the safe side I will put in a similar fix for the SWE.

Reply:
Now that I have done some testing including running one of the lake at rest simulations up to a final time of 500.0 there is no appreciable change in the bottom variable but, as I mentioned above, I still implemented a special version of the LLF dissipation and FluxHLL to zero out this dissipation entry.

Reply:
> no appreciable change in the bottom variable

Like "in the eyeball norm" or "when computing discrete errors"?

Reply:
the discrete errors change between the original FluxHLL and the new one was on the order of 3e-14 similar for the LLF

Reply:
Also in the case of a discontinuous bottom topography? That's the only case where I would expect to see an influence (since the jump of the bottom topography is zero otherwise whenever these fluxes are used).

Reply:
For discontinuous bottom topographies the `flux_hll` is not well-balanced. It would need modified with some kind of hydrostatic reconstruction. The `flux_lax_friedrichs` should be well-balanced which has revealed some weird behaviour. If I use the standard "standard" LLF the simulation is still well balanced whereas the modified LLF is not well-balanced and crashes.

Reply:
This is weird. The whole point of the modification is to make sure that the bottom topography does not change over time. If this crashes, there must be an issue elsewhere, doesn't it? Which "base flux" did you use - and did this base flux also add zeros for the bottom topography?

Reply:
Yeah, the real problem is that you shouldn't use the classic `flux_hll` or `flux_lax_friedrichs` for discontinuous bottom topography problems. They need some extra terms.

Reply:
Yeah, you need something like a hydrostatic reconstruction - could you please add this to the SWE TODO list, @andrewwinters5000?